src/repr/src/fixed_length.rs

```diff
@@ -0,0 +1,1005 @@
+use crate::row::{
+    date_to_array, naive_date_to_arrays, read_byte_array, read_date, read_naive_date, read_time,
+    time_to_arrays,
+};
+use crate::{ColumnType, Datum, Row, ScalarType, Timestamp};
+
+/// A typedef to ease use of 9-byte fixed-length representations.
+pub type Bytes9 = BytesN<9>;
+
+/// A helper trait to get references to `Row` based on type information that only manifests
+/// at runtime (typically originating from inferred schemas).
+pub trait IntoRowByTypes<'a>: Sized {
+    /// Obtains a reference to `Row` for an instance of `Self`, given a `Row` buffer and
+    /// a schema provided by `types`.
+    ///
+    /// Implementations are free to not use `row_buf` if a zero-copy implementation
+    /// can be provided. If the Row buffer is used, then the reference returned is to it.
+    /// Implementations are also free to place specific requirements on the given schema.
+    fn into_row(&'a self, row_buf: &'a mut Row, types: &[ColumnType]) -> &'a Row;
+}
+
+// Blanket identity implementation for Row.
+impl<'a> IntoRowByTypes<'a> for Row {
+    /// Performs a zero-copy reference forwarding without employing the `Row` buffer.
+    ///
+    /// This implementation panics if non-empty `types` are provided. This is because
+    /// `Row` is already self-describing and can use variable-length types, so we are
+    /// explicitly not validating the given schema.
+    fn into_row(&'a self, _row_buf: &'a mut Row, types: &[ColumnType]) -> &'a Row {
+        assert!(types.is_empty());
+        self
+    }
+}
+
+/// A helper trait to construct target values from input `Row` instances based on type
+/// information that only manifests at runtime (typically originating from inferred schemas).
+/// Typically, the target type will be of fixed-length without tags per column or `Row` itself.
+pub trait FromRowByTypes: Sized {
+    /// Obtains an instance of `Self` given an instance of `Row` and a schema provided
+    /// by `types`.
+    /// Implementations are free to place specific requirements on the given schema.
+    fn from_row(row: Row, types: &[ColumnType]) -> Self {
+        let iter = row.iter();
+        Self::from_datum_iter(iter, types)
+    }
+
+    /// Obtains an instance of `Self' given an iterator of borrowed datums and a schema
+    /// provided by `types`.
+    /// Implementations are free to place specific requirements on the given schema.
+    fn from_datum_iter<'a, I, D>(datum_iter: I, types: &[ColumnType]) -> Self
+    where
+        I: IntoIterator<Item = D>,
+        D: Borrow<Datum<'a>>;
+}
+
+impl FromRowByTypes for Row {
+    /// Returns the provided row itself.
+    ///
+    /// This implementation panics if non-empty `types` are provided. This is because
+    /// `Row` is already self-describing and can use variable-length types, so we are
+    /// explicitly not validating the given schema.
+    fn from_row(row: Row, types: &[ColumnType]) -> Self {
+        assert!(types.is_empty());
```

Comment:
Should we change the syntax to `types: Option<&[ColumnType]>` to be able to distinguish no columns from no types here, like you did elsewhere?

Reply:
It's not as critical to use an `Option` here as in when capturing the types, but it also does not hurt and might aid understandability. So I just went ahead and changed it.